src/nanoarrow/ipc/encoder.c

```diff
@@ -0,0 +1,108 @@
+// Can compile with -DNANOARROW_IPC_USE_STDATOMIC=0 or 1 to override
+// automatic detection
+#if !defined(NANOARROW_IPC_USE_STDATOMIC)
+#define NANOARROW_IPC_USE_STDATOMIC 0
+
+// Check for C11
+#if defined(__STDC_VERSION__) && __STDC_VERSION__ >= 201112L
+
+// Check for GCC 4.8, which doesn't include stdatomic.h but does
+// not define __STDC_NO_ATOMICS__
+#if defined(__clang__) || !defined(__GNUC__) || __GNUC__ >= 5
+
+#if !defined(__STDC_NO_ATOMICS__)
+#include <stdatomic.h>
+#undef NANOARROW_IPC_USE_STDATOMIC
+#define NANOARROW_IPC_USE_STDATOMIC 1
+#endif
+#endif
+#endif
+
+#endif
+
+#include "nanoarrow/ipc/flatcc_generated.h"
+#include "nanoarrow/nanoarrow.h"
+#include "nanoarrow/nanoarrow_ipc.h"
+
+// R 3.6 / Windows builds on a very old toolchain that does not define ENODATA
+#if defined(_WIN32) && !defined(_MSC_VER) && !defined(ENODATA)
+#define ENODATA 120
+#endif
+
+#define ns(x) FLATBUFFERS_WRAP_NAMESPACE(org_apache_arrow_flatbuf, x)
+
+#define FLATCC_RETURN_UNLESS_0(x) \
+  if (ns(x) != 0) return ENOMEM;
+
+struct ArrowIpcEncoderPrivate {
+  flatcc_builder_t builder;
+  struct ArrowBuffer buffers, nodes;
+};
+
+ArrowErrorCode ArrowIpcEncoderInit(struct ArrowIpcEncoder* encoder) {
+  NANOARROW_DCHECK(encoder);
+  memset(encoder, 0, sizeof(struct ArrowIpcEncoder));
+  encoder->encode_buffer = NULL;
+  encoder->encode_buffer_state = NULL;
+  encoder->codec = NANOARROW_IPC_COMPRESSION_TYPE_NONE;
+  encoder->private_data = ArrowMalloc(sizeof(struct ArrowIpcEncoderPrivate));
+  struct ArrowIpcEncoderPrivate* private =
+      (struct ArrowIpcEncoderPrivate*)encoder->private_data;
+  ArrowBufferInit(&private->buffers);
+  ArrowBufferInit(&private->nodes);
+  if (flatcc_builder_init(&private->builder) == -1) {
+    return ESPIPE;
+  }
+  return NANOARROW_OK;
+}
+
+void ArrowIpcEncoderReset(struct ArrowIpcEncoder* encoder) {
+  NANOARROW_DCHECK(encoder && encoder->private_data);
+  struct ArrowIpcEncoderPrivate* private =
+      (struct ArrowIpcEncoderPrivate*)encoder->private_data;
+  flatcc_builder_clear(&private->builder);
+  ArrowBufferReset(&private->nodes);
+  ArrowBufferReset(&private->buffers);
+  ArrowFree(private);
+  memset(encoder, 0, sizeof(struct ArrowIpcEncoder));
+}
+
+ArrowErrorCode ArrowIpcEncoderFinalizeBuffer(struct ArrowIpcEncoder* encoder,
+                                             struct ArrowBuffer* out) {
+  NANOARROW_DCHECK(encoder && encoder->private_data && out);
+  struct ArrowIpcEncoderPrivate* private =
+      (struct ArrowIpcEncoderPrivate*)encoder->private_data;
+  ArrowBufferReset(out);
+  size_t size = flatcc_builder_get_buffer_size(&private->builder);
+  if (size == 0) {
+    // Finalizing an empty flatcc_builder_t triggers an assertion
+    return NANOARROW_OK;
+  }
+
+  out->size_bytes = out->capacity_bytes = (int64_t)size;
+  out->data = (uint8_t*)flatcc_builder_finalize_buffer(&private->builder, &size);
+  return out->data ? NANOARROW_OK : ENOMEM;
```

Comment:
This depends. In order of increasing complexity:

- If it's acceptable to always copy from the buffer where flatcc has been operating
  in-place to a fresh output buffer then that's certainly simplest. In that case we
  resize an initialized `out` then memcpy.
- We could keep using the flatcc default allocator, but overwrite the output
  buffer's allocator to use `flatcc_builder_{free,alloc}`. (In this case, `out` is
  initialized when finalizing the encoder so the argument shouldn't be initialized)
  with the flatcc allocator and builder's
  result. This would save us a copy (unless we need to copy the result into a buffer
  from some other pool).
- We could eventually implement custom allocators for flatcc, which allows us to
  control what memory flatcc is using to build *and* saves us a finalizing copy
  no matter where we want the output. In this case again we require an externally
  initialized `out` whose allocator we reuse for flatcc building.

Reply:
I'll just copy for now